iothub/device/src/Transport/AmqpIoT/AmqpIoTReceivingLink.cs

```diff
@@ -258,88 +258,104 @@ private void DisposeDelivery(AmqpMessage amqpMessage, bool settled, Accepted acc
         internal void RegisterTwinListener(Action<Twin, string, TwinCollection> onDesiredPropertyReceived)
         {
             _onDesiredPropertyReceived = onDesiredPropertyReceived;
-            _receivingAmqpLink.RegisterMessageListener(OnDesiredPropertyReceived);
+            _receivingAmqpLink.RegisterMessageListener(OnTwinChangesReceived);
         }
 
-        private void OnDesiredPropertyReceived(AmqpMessage amqpMessage)
+        private void OnTwinChangesReceived(AmqpMessage amqpMessage)
         {
             if (Logging.IsEnabled)
             {
-                Logging.Enter(this, amqpMessage, $"{nameof(OnDesiredPropertyReceived)}");
+                Logging.Enter(this, amqpMessage, $"{nameof(OnTwinChangesReceived)}");
             }
 
             try
             {
                 _receivingAmqpLink.DisposeDelivery(amqpMessage, true, AmqpIoTConstants.AcceptedOutcome);
                 string correlationId = amqpMessage.Properties?.CorrelationId?.ToString();
+                int status = getStatus(amqpMessage);
 
-                if (!VerifyResponseMessage(amqpMessage))
+                if (status >= 400)
                 {
+                    // Handle failures
                     _onDesiredPropertyReceived.Invoke(null, correlationId, null);
-                }
-
-                Twin twin = null;
-                TwinCollection twinProperties = null;
+                    string error = null;
+                    using (StreamReader reader = new StreamReader(amqpMessage.BodyStream, System.Text.Encoding.UTF8))
+                    {
+                        error = reader.ReadToEnd();
+                    }
 
-                if (correlationId != null)
+                    if (status >= 500 || status == 429 || status == 408)
```

Comment:
again, extremely nit picky - this might fit into a helper method of sorts, or document which scenarios we'd like to retry for.

Reply:
What would you like the comment to be if the error codes are not sufficient?

Reply:
added this - Retry for Http status code request timeout, Too many requests and server errors